### PR TITLE
feat/add support for node os auto upgrade

### PR DESCRIPTION
**Pull request by bmiguel-teixeira**

### CODE REVIEW

File: internal/services/containers/kubernetes_cluster_resource.go [hunk 1]
```diff
@@ -160,6 +160,17 @@ func resourceKubernetesCluster() *pluginsdk.Resource {
...
+			"node_os_channel_upgrade": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(managedclusters.NodeOSUpgradeChannelNodeImage),
+					string(managedclusters.NodeOSUpgradeChannelSecurityPatch),
+					string(managedclusters.NodeOSUpgradeChannelUnmanaged),
+					string(managedclusters.NodeOSUpgradeChannelNone),
+				}, false),
+			},
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_resource.go:172]: The properties in the schema are ordered alphabetically, so this will need to be moved down.

File: internal/services/containers/kubernetes_cluster_resource.go [hunk 2]
```diff
@@ -1467,6 +1478,9 @@ func resourceKubernetesClusterCreate(d *pluginsdk.ResourceData, meta interface{}
 			UpgradeChannel: utils.ToPtr(managedclusters.UpgradeChannelNone),
 		}
 	}
+	if v, ok := d.GetOk("node_os_channel_upgrade"); ok {
+		parameters.Properties.AutoUpgradeProfile.NodeOSUpgradeChannel = utils.ToPtr(managedclusters.NodeOSUpgradeChannel(v.(string)))
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_resource.go:1482]: We should use the new pointer package for this 
```suggestion
		parameters.Properties.AutoUpgradeProfile.NodeOSUpgradeChannel = pointer.To(managedclusters.NodeOSUpgradeChannel(v.(string)))
```

File: internal/services/containers/kubernetes_cluster_resource.go [hunk 3]
```diff
@@ -160,6 +160,17 @@ func resourceKubernetesCluster() *pluginsdk.Resource {
...
 			},
 
+			"node_os_channel_upgrade": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(managedclusters.NodeOSUpgradeChannelNodeImage),
+					string(managedclusters.NodeOSUpgradeChannelSecurityPatch),
+					string(managedclusters.NodeOSUpgradeChannelUnmanaged),
+					string(managedclusters.NodeOSUpgradeChannelNone),
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_resource.go:170]: We don't typically expose values like `None` in the provider - some docs about that [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/schema-design-considerations.md#the-none-value-or-similar). This should default to `None` if the property has been removed after having previously been set.
```suggestion
```

File: internal/services/containers/migration/kubernetes_cluster.go [hunk 4]
```diff
@@ -1140,6 +1140,11 @@ func (k KubernetesClusterV0ToV1) Schema() map[string]*pluginsdk.Schema {
 			Optional: true,
 		},
 
+		"node_os_channel_upgrade": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+
```
Comment (stephybun) [inline internal/services/containers/migration/kubernetes_cluster.go:1147]: The schemas in migration are point in time references, so we don't need to add new properties here.
```suggestion

```

File: internal/services/containers/migration/kubernetes_cluster.go [hunk 5]
```diff
@@ -2421,6 +2426,11 @@ func (k KubernetesClusterV1ToV2) Schema() map[string]*pluginsdk.Schema {
...
```
Comment (stephybun) [inline internal/services/containers/migration/kubernetes_cluster.go:2433]: Same here.
```suggestion

```

File: website/docs/r/kubernetes_cluster.html.markdown [hunk 6]
```diff
@@ -85,6 +85,8 @@ In addition, one of either `identity` or `service_principal` blocks must be spec
 
 * `automatic_channel_upgrade` - (Optional) The upgrade channel for this Kubernetes Cluster. Possible values are `patch`, `rapid`, `node-image` and `stable`. Omitting this field sets this value to `none`.
 
+* `node_os_channel_upgrade` - (Optional) The upgrade channel for this Kubernetes Cluster Nodes OS Image. Possible values are `Unmanaged`, `SecurityPatch`, and `NodeImage`. Omitting this field sets this value to `none`.
+
```
Comment (stephybun) [inline website/docs/r/kubernetes_cluster.html.markdown:89]: Properties in the docs are ordered alphabetically as well so this will need to be moved down. Also since it's a preview feature could we add a note on that to inform users that it must be registered first before it can be set.
```suggestion
* `node_os_channel_upgrade` - (Optional) The upgrade channel for this Kubernetes Cluster Nodes OS Image. Possible values are `Unmanaged`, `SecurityPatch`, and `NodeImage`. Omitting this field sets this value to `none`.

-> **Note:** This requires that the Preview Feature `Microsoft.ContainerService/NodeOsUpgradeChannelPreview` is enabled and the Resource Provider is re-registered, see [the documentation](https://learn.microsoft.com/en-us/azure/aks/auto-upgrade-node-image#register-the-nodeosupgradechannelpreview-feature-flag) for more information.
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for this @bmiguel-teixeira. This is off to a great start. A few minor remarks but once those are resolved we can take another look through.

File: internal/services/containers/kubernetes_cluster_resource.go (hunk 1 above)
Comment (bmiguel-teixeira) [reply internal/services/containers/kubernetes_cluster_resource.go]: done.

File: internal/services/containers/kubernetes_cluster_resource.go (hunk 2 above)
Comment (bmiguel-teixeira) [reply internal/services/containers/kubernetes_cluster_resource.go]: done.

File: internal/services/containers/migration/kubernetes_cluster.go (hunk 4 above)
Comment (bmiguel-teixeira) [reply internal/services/containers/migration/kubernetes_cluster.go]: done.

File: internal/services/containers/migration/kubernetes_cluster.go (hunk 5 above)
Comment (bmiguel-teixeira) [reply internal/services/containers/migration/kubernetes_cluster.go]: done.

File: website/docs/r/kubernetes_cluster.html.markdown (hunk 6 above)
Comment (bmiguel-teixeira) [reply website/docs/r/kubernetes_cluster.html.markdown]: done.

**Review by stephybun — CHANGES_REQUESTED** [review]
Unfortunately we have quite a few of these test failures:
```
------- Stdout: -------
=== RUN   TestAccKubernetesCluster_addonProfileAciConnectorLinux
=== PAUSE TestAccKubernetesCluster_addonProfileAciConnectorLinux
=== CONT  TestAccKubernetesCluster_addonProfileAciConnectorLinux
------- Stderr: -------
panic: Invalid address to set: []string{"node_os_upgrade_channel"}
goroutine 631 [running]:
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*ResourceData).Set(0xc0037c9b80, {0x6f4cd0b, 0x17}, {0x6200460, 0xbf2c5a0})
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource_data.go:230 +0x291
github.com/hashicorp/terraform-provider-azurerm/internal/services/containers.resourceKubernetesClusterRead(0xc0037c9b80, {0x6364280?, 0xc000cee000})
  /opt/teamcity-agent/work/5e6516bb4d10eb66/internal/services/containers/kubernetes_cluster_resource.go:2226 +0xa74
github.com/hashicorp/terraform-provider-azurerm/internal/services/containers.resourceKubernetesClusterCreate(0x0?, {0x6364280?, 0xc000cee000})
  /opt/teamcity-agent/work/5e6516bb4d10eb66/internal/services/containers/kubernetes_cluster_resource.go:1560 +0x321a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x78a86c0?, {0x78a86c0?, 0xc001f1fb60?}, 0xd?, {0x6364280?, 0xc000cee000?})
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:695 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00449b5e0, {0x78a86c0, 0xc001f1fb60}, 0xc000f41ee0, 0xc0037c9a00, {0x6364280, 0xc000cee000})
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:837 +0xa85
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0002967e0, {0x78a86c0?, 0xc001f1faa0?}, 0xc000433770)
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/grpc_provider.go:1021 +0xe8d
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc00371dc20, {0x78a86c0?, 0xc001f1f3e0?}, 0xc000acc460)
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server/server.go:818 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x6affcc0?, 0xc00371dc20}, {0x78a86c0, 0xc001f1f3e0}, 0xc000acc3f0, 0x0)
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc001e301e0, {0x78b8140, 0xc0000f16c0}, 0xc001efb7a0, 0xc001eee630, 0xbee7800, 0x0)
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/google.golang.org/grpc/server.go:1340 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc001e301e0, {0x78b8140, 0xc0000f16c0}, 0xc001efb7a0, 0x0)
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/google.golang.org/grpc/server.go:1713 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/google.golang.org/grpc/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
  /opt/teamcity-agent/work/5e6516bb4d10eb66/vendor/google.golang.org/grpc/server.go:963 +0x28a
```

**Comment by bmiguel-teixeira** [comment]
hi @stephybun 

I had a typo that I missed. Already submitted the fix. Also ran a couple of acceptance tests w/ success. Take a look plz

File: internal/services/containers/kubernetes_cluster_other_resource_test.go [hunk 7]
```diff
@@ -499,6 +499,47 @@ func TestAccKubernetesCluster_upgradeChannel(t *testing.T) {
 	})
 }
 
+func TestAccKubernetesCluster_nodeOsUpgradeChannel(t *testing.T) {
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_other_resource_test.go:502]: We still have a test failure, it seems the casing on these needs to be updated - I also think the last step will cause an error when we pass in `"null"` as a string value, for the final step we need to use a config that omits the property entirely to test whether the provider correctly sets this to `none` when it's absent.

```
------- Stdout: -------
=== RUN   TestAccKubernetesCluster_nodeOsUpgradeChannel
=== PAUSE TestAccKubernetesCluster_nodeOsUpgradeChannel
=== CONT  TestAccKubernetesCluster_nodeOsUpgradeChannel
testcase.go:110: Step 1/8 error: Error running pre-apply refresh: exit status 1
Error: expected node_os_channel_upgrade to be one of [NodeImage SecurityPatch Unmanaged], got unmanaged
with azurerm_kubernetes_cluster.test,
on terraform_plugin_test.tf line 27, in resource "azurerm_kubernetes_cluster" "test":
27:   node_os_channel_upgrade   = "unmanaged"
--- FAIL: TestAccKubernetesCluster_nodeOsUpgradeChannel (2.06s)
FAIL
```

**Review by stephybun — CHANGES_REQUESTED** [review]

File: internal/services/containers/kubernetes_cluster_other_resource_test.go (hunk 7 above)
Comment (bmiguel-teixeira) [reply internal/services/containers/kubernetes_cluster_other_resource_test.go:503]: all tests are now passing. although im unsure if the approach is viable since we might be enforcing the NodeImage value. Can you take a look?

**Comment by stevehipwell** [comment]
What is actually pending on this PR, we really need to test this behaviour on our AKS clusters managed by Terraform.

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for your patience @bmiguel-teixeira.

Looking at the constraints and documented values for this field it looks like we will need to expose the `None` value since it is not the default value for the API. Apologies for pointing you in the wrong direction but looks like we will need to make an exception here.

Also I think the way to go will be to put validation in when setting `automatic_channel_upgrade` since the limitations for `node_os_channel_upgrade` only apply when `automatic_channel_upgrade` has been set to `node-image`. I've made the appropriate refactoring suggestions in-line (I couldn't use the suggestion feature everywhere since I can't make a suggestion on code that hasn't been touched).

Could you address the comments left in-line and also add a Note underneath the `automatic_channel_upgrade` property in the docs to the effect of:
```
-> **Note:** `node_os_channel_upgrade` must be set to `NodeImage` if `automatic_channel_upgrade` has been set to `node-image`
```

Once that's done we can take another look through this.

Thanks!

**Comment by stephybun** [comment]
@bmiguel-teixeira any updates on this? I would be happy to contribute those changes but am unable to since this is on your main branch.

**Comment by stephybun** [comment]
Superseded by #22187
